src/factory-reset/factory-reset-tool.c

```diff
@@ -0,0 +1,295 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse os-release: %m");
+
+        if (!id)
+                return log_error_errno(SYNTHETIC_ERRNO(EBADR), "os-release data lacks ID= field, refusing.");
+
+        sd_id128_t boot_id;
+        r = sd_id128_get_boot(&boot_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get boot ID: %m");
+
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_STRING("osReleaseId", id),
+                        SD_JSON_BUILD_PAIR_CONDITION(!!version_id, "osReleaseVersionId", SD_JSON_BUILD_STRING(version_id)),
+                        SD_JSON_BUILD_PAIR_CONDITION(!!image_id, "osReleaseImageId", SD_JSON_BUILD_STRING(image_id)),
+                        SD_JSON_BUILD_PAIR_CONDITION(!!image_version, "osReleaseImageVersion", SD_JSON_BUILD_STRING(image_version)),
+                        SD_JSON_BUILD_PAIR_ID128("bootId", boot_id));
+        if (r < 0)
+                return log_error_errno(r, "Failed to build JSON object: %m");
+
+        _cleanup_free_ char *formatted = NULL;
+        r = sd_json_variant_format(v, /* flags= */ 0, &formatted);
+        if (r < 0)
+                return log_error_errno(r, "Failed to format JSON object: %m");
+
+        r = efi_set_variable_string(EFI_SYSTEMD_VARIABLE_STR("FactoryResetRequest"), formatted);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set EFI variable FactoryResetRequest: %m");
+
+        log_debug("Set EFI variable FactoryResetRequest to '%s'.", formatted);
+
+        if (!arg_quiet)
+                log_info("Factory reset requested.");
+
+        return 0;
+}
+
+static int verb_cancel(int argc, char *argv[], void *userdata) {
+        int r;
+
+        FactoryResetMode f = factory_reset_mode();
+        if (f < 0)
+                return log_error_errno(f, "Failed to determine current factory reset mode: %m");
+        if (f == FACTORY_RESET_ON)
+                return log_error_errno(SYNTHETIC_ERRNO(EBUSY), "System already executing factory reset, cannot cancel.");
+        if (f != FACTORY_RESET_PENDING) {
+                log_info("No factory reset has been requested, cannot cancel, skipping.");
+                return 0;
+        }
+
+        if (!is_efi_boot()) {
+                if (!arg_quiet)
+                        log_info("Not an EFI boot, cannot remove FactoryResetMode EFI variable, not cancelling.");
+
+                return 0;
+        }
+
+        r = efi_set_variable(EFI_SYSTEMD_VARIABLE_STR("FactoryResetRequest"), /* value= */ NULL, /* size= */ 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to remove FactoryResetRequest EFI variable: %m");
+
+        if (!arg_quiet)
+                log_info("Factory reset cancelled.");
+
+        return 0;
+}
+
+static int retrigger_block_devices(void) {
```

Comment:
I think it is better to check with `udev_available()` earlier, and skip the operation on a container with a better log message.